feature/topic/src/main/kotlin/com/google/samples/apps/nowinandroid/feature/topic/TopicViewModel.kt

```diff
@@ -29,52 +27,74 @@ import com.google.samples.apps.nowinandroid.core.model.data.Topic
 import com.google.samples.apps.nowinandroid.core.model.data.UserNewsResource
 import com.google.samples.apps.nowinandroid.core.result.Result
 import com.google.samples.apps.nowinandroid.core.result.asResult
-import com.google.samples.apps.nowinandroid.feature.topic.navigation.TopicRoute
 import dagger.hilt.android.lifecycle.HiltViewModel
 import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
 import kotlinx.coroutines.flow.SharingStarted
 import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
 import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.filterNotNull
 import kotlinx.coroutines.flow.map
 import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.flow.update
 import kotlinx.coroutines.launch
 import javax.inject.Inject
 
 @HiltViewModel
 class TopicViewModel @Inject constructor(
-    savedStateHandle: SavedStateHandle,
     private val userDataRepository: UserDataRepository,
     topicsRepository: TopicsRepository,
-    userNewsResourceRepository: UserNewsResourceRepository,
+    userNewsResourceRepository: UserNewsResourceRepository
 ) : ViewModel() {
 
-    val topicId = savedStateHandle.toRoute<TopicRoute>().id
+    private val _topicId = MutableStateFlow<String?>(null)
+    val topicId = _topicId.asStateFlow()
+
+    private val _topicUIState = MutableStateFlow<TopicUiState>(TopicUiState.Loading)
+
+    private val _newsUIState = MutableStateFlow<NewsUiState>(NewsUiState.Loading)
+
 
-    val topicUiState: StateFlow<TopicUiState> = topicUiState(
-        topicId = topicId,
-        userDataRepository = userDataRepository,
-        topicsRepository = topicsRepository,
+    init {
+        viewModelScope.launch {
+            _topicId.filterNotNull().collect { topicId ->
+                combine(
+                    topicUiState(
+                        topicId = topicId,
+                        userDataRepository = userDataRepository,
+                        topicsRepository = topicsRepository,
+                    ),
+                    newsUiState(
+                        topicId = topicId,
+                        userDataRepository = userDataRepository,
+                        userNewsResourceRepository = userNewsResourceRepository,
+                    ),
+                ) { topicIUState, newsUIState ->
+                    _topicUIState.update { topicIUState }
+                    _newsUIState.update { newsUIState }
+                }.stateIn(viewModelScope)
```

Comment:
The `combine`s for old `topicId`s will continue to run right now, since these won't be cancelled if the topic id changes.